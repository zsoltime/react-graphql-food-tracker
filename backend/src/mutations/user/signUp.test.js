'use strict';

const signUp = require('./signUp');

describe('signUp mutation', () => {
  it('creates user and logs in', async () => {
    const args = { email: 'test@email.io', password: 'test1234' };
    const fakeUser = {
      id: 'testId',
      email: 'test@email.io',
      pendingEmail: null,
      permissions: ['USER'],
      profile: null,
    };
    const mockedContext = {
      db: {
        mutation: {
          createUser: jest.fn().mockImplementation(() => fakeUser),
        },
        query: {
          user: jest.fn().mockImplementation(() => null),
        },
      },
      response: {
        cookie: jest.fn(),
      },
    };

    const user = await signUp(null, args, mockedContext);

    expect(mockedContext.response.cookie).toHaveBeenCalled();
    expect(
      mockedContext.response.cookie.mock.calls[0]
    ).toMatchSnapshot();
    expect(user).toMatchSnapshot();
  });

  it('throws error if email is invalid', () => {
    const args = { email: 'invalidemail.io', password: 'test1234' };

    expect(signUp(null, args)).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws error if email is already in use', () => {
    const args = { email: 'zsolti@me.com', password: 'test1234' };
    const mockedContext = {
      db: {
        query: {
          user: jest
            .fn()
            .mockImplementation(() => ({ id: 'testUserId' })),
        },
      },
    };

    expect(
      signUp(null, args, mockedContext)
    ).rejects.toThrowErrorMatchingSnapshot();

    expect(mockedContext.db.query.user).toHaveBeenCalledWith({
      where: { email: args.email },
    });
  });

  it('throws error if password is invalid', () => {
    const args = { email: 'zsolti@me.com', password: 'invalid' };
    const mockedContext = {
      db: {
        query: {
          user: jest.fn().mockImplementation(() => null),
        },
      },
    };

    expect(
      signUp(null, args, mockedContext)
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
