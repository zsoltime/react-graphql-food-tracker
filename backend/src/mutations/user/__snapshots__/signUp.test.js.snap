// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signUp mutation creates user and logs in 1`] = `
Array [
  "token",
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VySWQiOiJ0ZXN0SWQiLCJpYXQiOjE1NDMyNDE2MDZ9.pvuxlFHMguOedl_DDC8bYZqTW7Bjku-xTiKDEryuiws",
  Object {
    "httpOnly": true,
    "maxAge": 7776000000,
    "secure": false,
  },
]
`;

exports[`signUp mutation creates user and logs in 2`] = `
Object {
  "email": "test@email.io",
  "id": "testId",
  "pendingEmail": null,
  "permissions": Array [
    "USER",
  ],
  "profile": null,
}
`;

exports[`signUp mutation throws error if email is already in use 1`] = `"Email is already in use"`;

exports[`signUp mutation throws error if email is invalid 1`] = `"Email is not valid"`;

exports[`signUp mutation throws error if password is invalid 1`] = `"Password must be at least 8 characters"`;
